Rewrite supported browsers summary, update Safari support info (#107)
stackblitz/docs: docs/platform/webcontainers/browser-support.md

_Last update: February 2023_

## Supported browsers

WebContainers are supported in all recent desktop browsers (fully in Chrome and Chrome-based browsers, in beta in Firefox and Safari), and partially supported on Android browsers.

### Desktop browsers

- Chrome: full support.
- Other Chromium-based browsers: full support. Some browsers may apply restrictive rules by default, and require specific configuration, like [Brave](#brave).
- Firefox: beta support ([see details](#firefox)).
- Safari: beta support since Safari 16.4, no support for earlier versions ([see details](#safari)).

### Mobile browsers

- Android: beta support for Chrome, Chromium-based browsers and Firefox. Depending on your device, large projects may run into memory limitations.

> - iOS and iPadOS: not supported yet. WebContainers can theoretically work on iOS 16.4 and later, but memory limitations for web apps in Safari for iOS and iPadOS are currently too strict to run most WebContainers-based projects.

Author: @d3lm I rewrote the info about iOS. Does that look good, or do we want to rework it a bit?

Reviewer: Yep I think that is accurate as of right now, at least until we test and possibly deploy a separate build for iPadOS.

## Web Platform requirements

StackBlitz requires some of the latest additions to the Web Platform to work correctly when running WebContainers-based projects. Most important among them are **`SharedArrayBuffer`** and **cross-origin isolation**.

### Cross-origin isolation

> [`SharedArrayBuffer`s](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer) (SABs) allow simultaneous access to a chunk of memory from multiple different workers. This is a powerful feature that was [disabled temporarily](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer#security_requirements) in light of potential security issues.

Author: Not sure this works, because the way it was addressed was by adding the COOP/COEP requirements. So they're still disabled by defaults and will always be.

> Cross-origin isolation is the key to enabling SABs: by properly configuring some of the headers and controlling which resources are served to browsers, a site can be considered [`crossOriginIsolated`](https://developer.mozilla.org/en-US/docs/Web/API/crossOriginIsolated) or, in other words, secure enough to use SABs. Both features are enabled in Chromium-based browsers (Chrome, Brave, Edge, etc.), Safari 16.4, and Firefox.

Reviewer 2: ```suggestion
Cross-origin isolation is the key to enabling SABs: by properly configuring some of the headers and controlling which resources are served to browsers, a site can be considered [`crossOriginIsolated`](https://developer.mozilla.org/en-US/docs/Web/API/crossOriginIsolated) or, in other words, secure enough to use SABs. Both features are enabled in Chromium-based browsers (Chrome, Brave, Edge, among others), Safari 16.4, and Firefox.
```

removing latinisms

_Last update: February 2023_

> ## Supported browsers

Reviewer 2: It definitely is a good title.

## Supported browsers

:::warning Note
There is a reported Chrome regression on Macbooks with M1 chip, which also affects the speed of some larger projects on WebContainers. Learn more about this issue in these bug reports: [issue 1228686](https://bugs.chromium.org/p/chromium/issues/detail?id=1228686) and [issue 1356099](https://bugs.chromium.org/p/chromium/issues/detail?id=1356099).

> :::

Author: I have a separate PR for removing the M1 warnings (https://github.com/stackblitz/docs/pull/106), but wanted to see the result of these page updates without those warnings too.

- Chrome: full support.
- Other Chromium-based browsers: full support. Some browsers may apply restrictive rules by default, and require specific configuration, like [Brave](#brave).
- Firefox: beta support ([see details](#firefox)).
- Safari: beta support since Safari 16.4, no support for earlier versions ([see details](#safari)).

### Mobile browsers

- Chrome, Chromium-based browsers and Firefox on Android: beta support. Depending on your device, large projects may run into memory limitations.

> - Safari on iOS 16.4: experimental support. Many projects may run into memory limitations and other compatibility issues.

Reviewer: Not yet supported though. I mean the editor will launch but WebContainer won't boot until we have a separate build for iOS and iPadOS. Maybe it also makes sense to separate iOS and iPadOS as there is overlap but the memory restrictions are different between both and tablets have a little higher memory limits. iOS will be a lot more restrictive and I don't think iOS is something we should support because almost no project with dependencies will run, and it would also mean we'd need a build for iOS and another one for iPadOS. Plus I think we can't even distinguish between the two via the user agent, can we?

Reviewer 3: we can distinguish via user agents as ipad gives a desktop UA. IOS gives a mobile UA. We can check for touch support to differentiate between desktop safari and ipad safari

Reviewer: Yea we may need to be a little more clever if it's possible to detect. I haven't looked into it but if you say we can then that'd be good. But we do want to mention the differences in memory restrictions between iPad and iOS.

## Web Platform requirements

StackBlitz requires some of the latest additions to the Web Platform to work correctly when running WebContainers-based projects. Most important among them are **`SharedArrayBuffer`** and **cross-origin isolation**.

### Cross-origin isolation

[`SharedArrayBuffer`s](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer) (SABs) allow simultaneous access to a chunk of memory from multiple different workers. This is a powerful feature that was [disabled temporarily](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer#security_requirements) in light of potential security issues.

Cross-origin isolation is the key to enabling SABs: by properly configuring some of the headers and controlling which resources are served to browsers, a site can be considered [`crossOriginIsolated`](https://developer.mozilla.org/en-US/docs/Web/API/crossOriginIsolated) or, in other words, secure enough to use SABs. Both features are enabled in Chromium-based browsers (Chrome, Brave, Edge, etc.), Safari 16.4, and Firefox.

> However, for cross-origin isolation to work for our use case, you need to be able to embed arbitrary resources: to be able to write and test your web application seamlessly, regardless of which images or scripts you choose to include. For this to work, a [new `credentialless` mode](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Embedder-Policy#credentialless) of cross-origin isolation that allows this is needed.

Author: Changed the link to `credentialless` from the WICG repo (https://github.com/WICG/credentiallessness) to a MDN page, because the proposal was merged in the HTML and Fetch specifications.

MDN doesn't have a lot of places or details we can link to, this was the best fit I could find at this time. (There are pages about `<iframe credentialless>` and `window.credentialless` but those are different features.)

We have beta support for Firefox. Please try it and [share feedback with us](https://github.com/stackblitz/webcontainer-core/issues/new/choose)!

Firefox does not fully support the required mode for cross-origin isolation, so you might encounter limitations when **running a server in a preview frame** within the StackBlitz editor. Third-party assets might get blocked due to the limitations of cross-origin isolation policies. However, you can work around this by opening your server preview in a separate window.

> Because Firefox and Node.js use different JavaScript engines (SpiderMonkey and V8), there is a small risk of running into [JS engine differences](#js-engine-differences).

Reviewer 2: ```suggestion
Because Firefox and Node.js use different JavaScript engines (SpiderMonkey and V8, respectively), there is a small risk of running into [JS engine differences](#js-engine-differences).
```

## Safari

We have beta support for Safari on macOS, starting with Safari 16.4. Please try it and [report issues here](https://github.com/stackblitz/webcontainer-core/issues/new/choose).

Safari does not fully support the required mode for cross-origin isolation, so you might encounter limitations when **running a server in a preview frame** within the StackBlitz editor. Third-party assets might get blocked due to the limitations of cross-origin isolation policies. However, you can work around this by opening your server preview in a separate window.

> Because Safari and Node.js use different JavaScript engines (JavaScriptCore and V8), there is a small risk of running into [JS engine differences](#js-engine-differences).

Reviewer 2: ```suggestion
Because Safari and Node.js use different JavaScript engines (JavaScriptCore and V8, respectively), there is a small risk of running into [JS engine differences](#js-engine-differences).
```

Because Firefox and Node.js use different JavaScript engines (SpiderMonkey and V8), there is a small risk of running into [JS engine differences](#js-engine-differences).

## Safari

We have beta support for Safari on macOS, starting with Safari 16.4. Please try it and [report issues here](https://github.com/stackblitz/webcontainer-core/issues/new/choose).

Safari does not fully support the required mode for cross-origin isolation, so you might encounter limitations when **running a server in a preview frame** within the StackBlitz editor. Third-party assets might get blocked due to the limitations of cross-origin isolation policies. However, you can work around this by opening your server preview in a separate window.

Because Safari and Node.js use different JavaScript engines (JavaScriptCore and V8), there is a small risk of running into [JS engine differences](#js-engine-differences).

> Older Safari versions are not suported. They lack a few necessary Web Platform features, such as [`Atomics.waitAsync`](https://github.com/tc39/proposal-atomics-wait-async) and [lookbehind in regular expressions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Assertions), which cannot be polyfilled.

Author: Rewrote the Safari section.

- Structure and language is similar to the Firefox section.
- Minimized the information about features missing in Safari (now that it's about older versions of Safari, we can de-emphasize it).

Reviewer 2: awesome
